### ISSUE TITLE
Subscriptions getting On hold for 26 hours for INR card payments before charge

### ISSUE THREAD

**Issue by ramshengale** [woocommerce/woocommerce-gateway-stripe#3177]
**Describe the bug**
This bug affects all Indian credit card holders for autopay mandates with Woocommerce Subscriptions.

Please refer to [this document](https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents&lang=php#predebit-notification). 

When there is a renewal, the subscription status changes to **On-hold** and Woocommerce sends payment charge request to Stripe. But due to the regulation the payment_intent status changes to "processing" and the charge is held for 26 hours. 

Now I understand that this is required for regulatory compliance but **this is causing issue for our customers as their subscriptions stays On-hold for 26 hours straight**. When they try to pay manually, they're not able to do that as well.

Sometimes they even end up buying a new subscription out of confusion.

**To Reproduce**
Buy subscription with an Indian card. That'll setup auto mandate. During renewal, if the charge has to happen on say 1st July, 2024 at 00:00, the subscription will turn On-hold and will stay like that for 26 hours. Then the actual charge will happen at 2nd July 02:00 (after 26 hours) and the subscription will turn Active. During this 26 hours period the customer won't be able to use the subscription (because it'll be On-hold) and also won't be able to manually pay (because it's not allowed when payment_intent status = pending).

**Expected behavior**
The renewal should happen immediately at 1st July, 2024 at 00:00 in case processing.card.customer_notification.approval_requested = false or the order status should not change to On-hold until actual charge happens.

**Possible solutions**
1. Do not keep the 26 hour period when processing.card.customer_notification.approval_requested = false.
2. If 26 hours are required, send the charge request from WooCommerce 26 hours before the renewal date/time and keep subscription active during that time.
3. Do not put the order On-hold when there is autopay mandate for Indian cards.

**Comment by ramshengale** [comment]
Can someone look into this?

**Comment by ramshengale** [comment]
Bumping this up because there has been no response to this since I created this.

**Comment by ramshengale** [comment]
A simple quick fix for this would be to add `do_action( 'wc_gateway_stripe_must_authorize_off_session', $response, $renewal_order );` at the bottom of the following **elseif** condition:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/55daf177b13281752e1dbd445d9e2ffad3763f8b/includes/compat/trait-wc-stripe-subscriptions.php#L379-L395

That way we will be able to hook into it and change the subscription status to Active temporarily and resolve this issue until a more robust solution is found.

Can this be implemented please?

**Comment by ramshengale** [comment]
When are you going to look into this issue?

**Comment by hsingyuc** [comment]
Hi @ramshengale, thank you for reporting this issue, and I apologize for the delayed response. We'll investigate the issue soon.

**Comment by hsingyuc** [comment]
[Duplicate issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3354).

[More context on Slack](https://a8c.slack.com/archives/C05GDK92WGM/p1741115739910029?thread_ts=1741106789.317659&cid=C05GDK92WGM).

**Comment by hsingyuc** [comment]
I'm waiting on Stripe's response [here](https://a8c.slack.com/archives/C9976E5MJ/p1741618651375549).

**Comment by ramshengale** [comment]
Thanks for taking this up @hsingyuc. Do ask your team to go through the suggestions I've provided in that [duplicate issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3354).

For the time being, a simple temporary solution for would be to just provide a filter / action hook at the end of this elseif condition:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L540-L556

That way we can manage the logic at our end. Then when you implement an actual solution to this, we'll remove our custom code. 

Can this be done?

**Comment by ramshengale** [comment]
On some digging into this, I figured out the exact solution for this. Kindly review this and also share your dev team as this should resolve this issue.

In the code below, we are not checking the condition if **approval_requested** is **false** or **true**. 

```
protected function must_authorize_off_session( $payment_intent ) {
		return ! empty( $payment_intent->status )
			&& WC_Stripe_Intent_Status::PROCESSING === $payment_intent->status
			&& ! empty( $payment_intent->processing->card->customer_notification->completes_at );
	}
```

As per Stripe's [documentation](https://docs.stripe.com/india-recurring-payments?integration=paymentIntents-setupIntents&lang=php#predebit-notification), in case `approval_requested = true`, that means the payment amount is over ₹15,000 and we need to wait for 26 hours for user to approve the payment. In that case the plugin can continue to work as it is by putting the order on hold until the charge happens after 26 hours.

But in case `approval_requested = false`, that means the payment amount can be charged immediately and there's no need to wait for 26 hours. Replacing the above code with the code should resolve this at least for payments that are under ₹15,000:

```
protected function must_authorize_off_session( $payment_intent ) {
	return ! empty( $payment_intent->status )
		&& WC_Stripe_Intent_Status::PROCESSING === $payment_intent->status
		&& ! empty( $payment_intent->processing->card->customer_notification )
		&& isset( $payment_intent->processing->card->customer_notification->approval_requested )
		&& true === $payment_intent->processing->card->customer_notification->approval_requested;
}

```

Kindly review let me know if you can implement this.

**Comment by MrRedFlannel** [comment]
It has been noted that users who switch subscriptions to lower plans are receiving an error:

![Image](https://github.com/user-attachments/assets/b9235296-4349-4dab-94e4-76636366b742)

This is related to the autopay mandate, but it's unclear if the user has already given the mandate for ₹1887.06, any lower payment (₹1599.20) should go through without issues.

9464278-zd-a8c
cc @ramshengale

**Comment by annemirasol** [comment]
My thoughts on the issue and potential solutions:

1. The plugin itself isn't holding or delaying the charges in this case (INR recurring payments). When we send a payment intent to Stripe, we use "auto-confirm," which means we ask Stripe to charge the customer _immediately_. Whether the charge goes through right away or is delayed is completely up to Stripe.

2. When Stripe delays the charge attempt due to the pre-debit notification period, the plugin can only wait for the `payment_intent.succeeded` webhook event. This event usually arrives around the `completes_at` time, which is a future timestamp provided by Stripe.
   
   - The order note that says "The customer must authorize this payment..." shows that `completes_at` timestamp from Stripe.

3. Changing the plugin’s `WC_Stripe_Subscriptions_Trait::must_authorize_off_session()` method to return false, as suggested, won’t help. Even then, the Stripe response won’t include a charge ID (latest_charge is still null), because no charge has actually been made yet.
   
   - A better name for that function might be `is_charge_attempt_delayed()`, which would make it clearer that this is about Stripe delaying the charge, and not about needing customer authorization.

4. The most realistic solution I can see would be to add a new WordPress Action at the end of the delayed charge handling logic. This would let site owners decide whether to activate the subscription while waiting for the charge to succeed. (See PR #4251.)

**Comment by ramshengale** [comment]
I want to reopen this issue because there are still a couple of issues pending.

1. The issue that's mentioned by @MrRedFlannel still exists.
2. This solution give above by @annemirasol in point #4 worked out well using the following action function:

```
public static function handle_delayed_payment_charge( $payment, $renewal_order ) { 
		$subscriptions = wcs_get_subscriptions_for_renewal_order( $renewal_order->get_id() );
		foreach ( $subscriptions as $subscription ) {
			wc_get_logger()->debug( "Reactivating subscription ID #{$subscription->get_id()} as charges are likely to clear." );
            $subscription->add_order_note("Reactivating subscription ID #{$subscription->get_id()} as charges are likely to clear.");
			$subscription->update_status( 'active' );
		}
}
```
But now, after the renewal payment actually happens on Stripe (after 26 hours), the status of order is not getting updated to **Completed**. It stays **Pending Payment**. We pushed the latest code live a week ago and see that all renewal payments happened successfully but the plugin was not able to capture the webhook request for these payments. Hence all subscriptions are Active but their last orders are all showing **Pending Payment**.
